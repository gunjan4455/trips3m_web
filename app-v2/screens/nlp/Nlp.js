import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import config from 'appConfig';

import worker from './apiWorker';
import WebWorker from 'helpers/webWorker';
import BudgetBanner from 'modules/nlp/BudgetBanner';
import NurturingOption from 'modules/nlp/NurturingOption';
import { ThumbsUp, ThumbsDown } from 'helpers/Icon/Icon';
import { replaceUniqueAndSecretKeys } from 'helpers/urlHelpers';
import { EVENT_TYPES } from 'constants/nlp';
import { getPreferencesData, getPreferenceParams } from 'screens/nlp/action';
import { getTripDays } from 'screens/nlp/reducer';
import { getMaxRatingOfHotelAttributes } from 'helpers/nlp';
import * as nlpReducers from 'screens/nlp/reducerIndex';
import { rescue } from 'helpers/utils';
import { trackSegment } from 'actions/segmentEvents';
import { ORGANISM_VIEWED } from 'actions/segmentEvents';

export { getPreferencesData, getPreferenceParams, getTripDays, getMaxRatingOfHotelAttributes };
export { nlpReducers };

const NurturingOptionData = [
  {
    title: `I’m <span className="pfc5">okay</span> with this budget`,
    text: 'Continue with the current plan',
    subText: 'View Packages based on your preferences',
    Icon: ThumbsUp,
    redirection: true
  },
  {
    title: `I’m <span className="pfc5">not okay</span> with this budget`,
    text: 'Edit Preferences',
    subText: 'to match your budget',
    Icon: ThumbsDown
  }
];

class NurturingLandingPage extends PureComponent {

  state = { disabled: false };

  componentDidMount() {
    const { match: { params: { uniqueKey, secretKey } }} = this.props;
    this.worker = new WebWorker(worker);
    this.worker.onerror = (e) => {
      this.worker.postMessage({ type: 'CLEAR ALIVE STATUS' });
      this.worker.terminate();
    };
    const url = replaceUniqueAndSecretKeys(config.apiEndpoints.nuturePreferences.updateAlive.url,
      uniqueKey, secretKey);
    this.worker.postMessage({
      type: EVENT_TYPES.UPDATE_ALIVE,
      url: `${config.api_protocol}//${config.browsing.external.host}/${url}`,
      apiKey: config.api.key
    });
    this.trackSegmentEvent({
      event: ORGANISM_VIEWED
    });
  }

  savePreferencesAndRedirect = (rtripParams) => {
    if(this.state.disabled) {
      return true;
    }
    const { rtripData, savePreferences } = this.props;
    const data = {
      isSubmit: true,
      redis_key: rtripData.uniq_key,
      requested_trip: rtripParams ? rtripParams : rtripData.rtrip_params,
      secret_key: rtripData.secret_key,
      nurtured: true
    };
    this.setState({ disabled: true });
    return savePreferences(data)
      .then((res) => {
        this.worker.postMessage({ type: EVENT_TYPES.CLEAR_ALIVE });
        this.worker.terminate();
        console.log('redirectUrl:: ', res.data.redirect_url);
        window.location.href = res.data.redirect_url;
      })
      .catch(() => {
        this.worker.postMessage({ type: EVENT_TYPES.CLEAR_ALIVE });
        this.worker.terminate();
      });
  };

  trackSegmentEvent = (data = {}) => {
    rescue(() => {
      trackSegment({
        event: data.event,
        section: data.section || '',
        object: data.object,
        cta: data.cta
      });
    });
  };

  componentWillUnmount() {
    this.worker.postMessage({ type: EVENT_TYPES.CLEAR_ALIVE });
    this.worker.terminate();
  }

  render() {
    const { preference } = this.props;
    return (
      <div>
        <BudgetBanner
          destination={preference.destinationName}
          priceMax={preference.priceMax}
          priceMin={preference.priceMin}
          savePreferencesAndRedirect={this.savePreferencesAndRedirect}
          adult={preference.adult}
          child={preference.child}
        />
        {NurturingOptionData.map((item, index) =>
          <NurturingOption
            {...item}
            hideBottomBorder={index === NurturingOptionData.length - 1}
            savePreferencesAndRedirect={this.savePreferencesAndRedirect}
            trackSegmentEvent={this.trackSegmentEvent}
            disabled={this.state.disabled}
            key={index}
            preference={preference}
          />)
        }
      </div>
    );
  }
}

NurturingLandingPage.propTypes = {
  preference: PropTypes.object.isRequired,
  rtripData: PropTypes.object.isRequired,
  savePreferences: PropTypes.func.isRequired,
  match: PropTypes.object.isRequired
};

export default NurturingLandingPage;
