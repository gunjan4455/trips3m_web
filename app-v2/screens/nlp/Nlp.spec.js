import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount, configure } from 'enzyme';

import Nlp from './Nlp.js';

configure({ adapter: new Adapter() });

const DEFAULT_PROPS = {
  "params": {
    "uniqueKey": "1582648880674_lzgfh-ur",
    "secretKey": "977886e355826ccf"
  },
  "preference": {
    "destinationId": 56,
    "destinationName": "Kerala",
    "destinationType": "domestic",
    "days": 3,
    "maxHotelRating": 4,
    "flights": false,
    "travelMonth": 2,
    "priceMin": 55152,
    "priceMax": 142767
  },
  "rtripData": {
    "rtrip_params": {
      "email": "asqwesd@gmail.com",
      "dont_utilize_budget_new": true,
      "budget_new": true,
      "start_date_noidea": false,
      "whatsapp_opt_status": true,
      "adult": "2",
      "child": "0",
      "isd_code": "+91",
      "child_age_answers_attributes": [
        {
          "user_input": "0",
          "customization_id": "7",
          "option_id": "29"
        }
      ],
      "adminComments": "  #Form_filled_on https://localhost:3000/ #trip_origin_product:public_mweb  ",
      "syncErrors": {},
      "to_loc": "Kerala",
      "from_loc": "Lane Number 7, Maruti Udyog, Sector 18, Gurugram, Haryana 122008, India",
      "start_date": "2020-02-20",
      "start_date_flexible": false,
      "trip_days": 3,
      "hotel_needed_answer_attributes": {
        "option_id": "95",
        "customization_id": "22"
      },
      "hotel_category_answers_attributes": [
        {
          "option_id": "7",
          "customization_id": "2"
        }
      ],
      "other_answers_attributes": [
        {
          "option_id": "712",
          "customization_id": "218"
        }
      ],
      "budget": "8250..9000",
      "infant": "0",
      "phone_no": "9123123123",
      "from_slider_form": "true",
      "create_your_own_id": "false",
      "new_form": "true",
      "referer": "direct",
      "addtl_info": " ",
      "attribution_attributes": {
        "trip_origin_product": "public_mweb",
        "trip_origin_uri": "public_mweb::F1 Page::::::",
        "context_page_url": "https://localhost:3000/",
        "session_uuid": "a40b511af1d21e8f13f6d2fb06df23162636e6f1:::12-1-120",
        "page_uuid": "aa49966360adbcaac04771c21170c82b1ffebb77:::dvlCoZd9CS3lVzwGAvPtyOlYUnSB8E",
        "landing_page_uuid": "e7753672cd47e8985e76f26e274745cd5f4e59a2:::eijZlfoJp385EUrS8PvHkE0sxpdERM",
        "referer_page_uuid": "e7753672cd47e8985e76f26e274745cd5f4e59a2:::eijZlfoJp385EUrS8PvHkE0sxpdERM",
        "funnel_uuid": "0Ienf4peUu8YzMQ3rnJ9"
      }
    },
    "referrer": {
      "__utmz": null,
      "__utmzz": "utmcsr=(direct)|utmccn=(direct)|utmcmd=(none)",
      "cookie_referer": null,
      "new_referer": null,
      "user_agent": "Mozilla/5.0 (Linux; Android 5.0; SM-G900P Build/LRX21T) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Mobile Safari/537.36"
    },
    "mapped_to_loc_id": 56,
    "mapped_to_loc_name": "Kerala",
    "mapped_to_type": "domestic",
    "mapped_from_location_id": 20910,
    "mapped_from_location_name": "Kupwara",
    "start_date_month": 2,
    "adults": 3,
    "redis_key": "0ienf4peuu8yzmq3rnj9",
    "secret_key": "8cb0822f788b351d"
  },
  savePreferences: () => {
  }
}

class WorkerStub {
  constructor(stringUrl) {
    this.url = stringUrl;
    this.onmessage = () => {
    };
  }

  postMessage(msg) {
    this.onmessage(msg);
  }
}

const shallowWrapper = (props) => {
  window.Worker = WorkerStub;
  window.URL.createObjectURL = jest.fn();
  return shallow(<Nlp {...props} />);
}

describe('Should render correctly', () => {
  let props = { ...DEFAULT_PROPS };

  it('should render correctly', () => {
    const wrapper = shallowWrapper(props);
    expect(wrapper).toMatchSnapshot();
  });
});
