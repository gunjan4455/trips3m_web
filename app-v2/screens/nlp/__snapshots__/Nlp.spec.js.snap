// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NurturingLandingPage
    params={
      Object {
        "secretKey": "977886e355826ccf",
        "uniqueKey": "1582648880674_lzgfh-ur",
      }
    }
    preference={
      Object {
        "days": 3,
        "destinationId": 56,
        "destinationName": "Kerala",
        "destinationType": "domestic",
        "flights": false,
        "maxHotelRating": 4,
        "priceMax": 142767,
        "priceMin": 55152,
        "travelMonth": 2,
      }
    }
    rtripData={
      Object {
        "adults": 3,
        "mapped_from_location_id": 20910,
        "mapped_from_location_name": "Kupwara",
        "mapped_to_loc_id": 56,
        "mapped_to_loc_name": "Kerala",
        "mapped_to_type": "domestic",
        "redis_key": "0ienf4peuu8yzmq3rnj9",
        "referrer": Object {
          "__utmz": null,
          "__utmzz": "utmcsr=(direct)|utmccn=(direct)|utmcmd=(none)",
          "cookie_referer": null,
          "new_referer": null,
          "user_agent": "Mozilla/5.0 (Linux; Android 5.0; SM-G900P Build/LRX21T) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.88 Mobile Safari/537.36",
        },
        "rtrip_params": Object {
          "addtl_info": " ",
          "adminComments": "  #Form_filled_on https://localhost:3000/ #trip_origin_product:public_mweb  ",
          "adult": "2",
          "attribution_attributes": Object {
            "context_page_url": "https://localhost:3000/",
            "funnel_uuid": "0Ienf4peUu8YzMQ3rnJ9",
            "landing_page_uuid": "e7753672cd47e8985e76f26e274745cd5f4e59a2:::eijZlfoJp385EUrS8PvHkE0sxpdERM",
            "page_uuid": "aa49966360adbcaac04771c21170c82b1ffebb77:::dvlCoZd9CS3lVzwGAvPtyOlYUnSB8E",
            "referer_page_uuid": "e7753672cd47e8985e76f26e274745cd5f4e59a2:::eijZlfoJp385EUrS8PvHkE0sxpdERM",
            "session_uuid": "a40b511af1d21e8f13f6d2fb06df23162636e6f1:::12-1-120",
            "trip_origin_product": "public_mweb",
            "trip_origin_uri": "public_mweb::F1 Page::::::",
          },
          "budget": "8250..9000",
          "budget_new": true,
          "child": "0",
          "child_age_answers_attributes": Array [
            Object {
              "customization_id": "7",
              "option_id": "29",
              "user_input": "0",
            },
          ],
          "create_your_own_id": "false",
          "dont_utilize_budget_new": true,
          "email": "asqwesd@gmail.com",
          "from_loc": "Lane Number 7, Maruti Udyog, Sector 18, Gurugram, Haryana 122008, India",
          "from_slider_form": "true",
          "hotel_category_answers_attributes": Array [
            Object {
              "customization_id": "2",
              "option_id": "7",
            },
          ],
          "hotel_needed_answer_attributes": Object {
            "customization_id": "22",
            "option_id": "95",
          },
          "infant": "0",
          "isd_code": "+91",
          "new_form": "true",
          "other_answers_attributes": Array [
            Object {
              "customization_id": "218",
              "option_id": "712",
            },
          ],
          "phone_no": "9123123123",
          "referer": "direct",
          "start_date": "2020-02-20",
          "start_date_flexible": false,
          "start_date_noidea": false,
          "syncErrors": Object {},
          "to_loc": "Kerala",
          "trip_days": 3,
          "whatsapp_opt_status": true,
        },
        "secret_key": "8cb0822f788b351d",
        "start_date_month": 2,
      }
    }
    savePreferences={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <BudgetBanner
          adult={1}
          child={0}
          destination="Kerala"
          priceMax={142767}
          priceMin={55152}
          savePreferencesAndRedirect={[Function]}
        />,
        Array [
          <NurturingOption
            Icon={[Function]}
            disabled={false}
            hideBottomBorder={false}
            redirection={true}
            savePreferencesAndRedirect={[Function]}
            subText="View Packages based on your preferences"
            text="Continue with the current plan"
            title="I’m <span className=\\"pfc5\\">okay</span> with this budget"
            trackSegmentEvent={[Function]}
          />,
          <NurturingOption
            Icon={[Function]}
            disabled={false}
            hideBottomBorder={true}
            redirection={false}
            savePreferencesAndRedirect={[Function]}
            subText="to match your budget"
            text="Edit Preferences"
            title="I’m <span className=\\"pfc5\\">not okay</span> with this budget"
            trackSegmentEvent={[Function]}
          />,
        ],
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "adult": 1,
          "child": 0,
          "destination": "Kerala",
          "priceMax": 142767,
          "priceMin": 55152,
          "savePreferencesAndRedirect": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "class",
        "props": Object {
          "Icon": [Function],
          "disabled": false,
          "hideBottomBorder": false,
          "redirection": true,
          "savePreferencesAndRedirect": [Function],
          "subText": "View Packages based on your preferences",
          "text": "Continue with the current plan",
          "title": "I’m <span className=\\"pfc5\\">okay</span> with this budget",
          "trackSegmentEvent": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "Icon": [Function],
          "disabled": false,
          "hideBottomBorder": true,
          "redirection": false,
          "savePreferencesAndRedirect": [Function],
          "subText": "to match your budget",
          "text": "Edit Preferences",
          "title": "I’m <span className=\\"pfc5\\">not okay</span> with this budget",
          "trackSegmentEvent": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <BudgetBanner
            adult={1}
            child={0}
            destination="Kerala"
            priceMax={142767}
            priceMin={55152}
            savePreferencesAndRedirect={[Function]}
          />,
          Array [
            <NurturingOption
              Icon={[Function]}
              disabled={false}
              hideBottomBorder={false}
              redirection={true}
              savePreferencesAndRedirect={[Function]}
              subText="View Packages based on your preferences"
              text="Continue with the current plan"
              title="I’m <span className=\\"pfc5\\">okay</span> with this budget"
              trackSegmentEvent={[Function]}
            />,
            <NurturingOption
              Icon={[Function]}
              disabled={false}
              hideBottomBorder={true}
              redirection={false}
              savePreferencesAndRedirect={[Function]}
              subText="to match your budget"
              text="Edit Preferences"
              title="I’m <span className=\\"pfc5\\">not okay</span> with this budget"
              trackSegmentEvent={[Function]}
            />,
          ],
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "adult": 1,
            "child": 0,
            "destination": "Kerala",
            "priceMax": 142767,
            "priceMin": 55152,
            "savePreferencesAndRedirect": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "Icon": [Function],
            "disabled": false,
            "hideBottomBorder": false,
            "redirection": true,
            "savePreferencesAndRedirect": [Function],
            "subText": "View Packages based on your preferences",
            "text": "Continue with the current plan",
            "title": "I’m <span className=\\"pfc5\\">okay</span> with this budget",
            "trackSegmentEvent": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "Icon": [Function],
            "disabled": false,
            "hideBottomBorder": true,
            "redirection": false,
            "savePreferencesAndRedirect": [Function],
            "subText": "to match your budget",
            "text": "Edit Preferences",
            "title": "I’m <span className=\\"pfc5\\">not okay</span> with this budget",
            "trackSegmentEvent": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
